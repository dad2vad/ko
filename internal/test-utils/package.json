{
  "name": "@kosko/test-utils",
  "version": "0.0.0",
  "private": true,
  "engines": {
    "node": ">=14.18.0"
  },
  "main": "dist/index.node.cjs",
  "module": "dist/index.node.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "build-scripts"
  },
  "exports": {
    "types": "./dist/index.d.ts",
    "deno": "./dist/index.deno.mjs",
    "node": {
      "import": "./dist/index.node.mjs",
      "require": "./dist/index.node.cjs"
    },
    "import": "./dist/index.base.mjs",
    "require": "./dist/index.node.cjs"
  },
  "dependencies": {
    "execa": "^5.1.1",
    "temp-dir": "^2.0.0",
    "tmp": "^0.2.1",
    "tmp-promise": "^3.0.3"
  },
  "devDependencies": {
    "@kosko/build-scripts": "workspace:^",
    "@types/tmp": "^0.2.3"
  }
}
