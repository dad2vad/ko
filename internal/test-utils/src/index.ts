export * from "./getRejectedValue";
export * from "./installPackage";
export * from "./makeTempDir";
export * from "./makeTempFile";
export * from "./projectRoot";
