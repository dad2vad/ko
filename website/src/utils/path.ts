export const sep = "/";

export function isAbsolute(path: string): boolean {
  return path.startsWith(sep);
}

export function isRelative(path: string): boolean {
  return path.startsWith(`.${sep}`) || path.startsWith(`../${sep}`);
}

export function dirname(path: string): string {
  const index = path.lastIndexOf(sep);
  if (index === -1) return ".";
  return path.substring(0, index) || sep;
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf(sep);
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split(sep).filter(Boolean);
  const toParts = to.split(sep).filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join(sep);
}
