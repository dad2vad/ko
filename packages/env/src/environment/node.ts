import { getRequireExtensions } from "@kosko/require";
import { join } from "node:path";
import { formatPath } from "../paths";
import { toArray } from "@kosko/common-utils";
import {
  createEnvironment,
  createLoaderReducers,
  createReducerList
} from "./base";
import { Environment, Loader, ReducerExecutor, ReducerList } from "./types";
import { cwd as getCwd } from "node:process";

export function getNodeExtensions(): string[] {
  return getRequireExtensions().map((ext) => ext.substring(1));
}

/**
 * Options for creating a Node.js environment.
 *
 * @public
 */
export interface NodeEnvironmentOptions {
  /**
   * Current working directory (CWD).
   *
   * @defaultValue `process.cwd()`
   */
  cwd?: string;
}

export interface InternalNodeEnvironmentOptions extends NodeEnvironmentOptions {
  createReducerExecutor(this: void, reducers: ReducerList): ReducerExecutor;
  requireModule(this: void, environment: Environment, id: string): any;
  mergeValues(this: void, data: any[]): any;
}

export function createNodeEnvironment({
  cwd = getCwd(),
  createReducerExecutor,
  requireModule,
  mergeValues
}: InternalNodeEnvironmentOptions): Environment {
  function requireAllEnvs(template: string, component?: string) {
    if (!environment.env) return [];

    const envs = toArray(environment.env);

    return envs.map((env) => {
      const path = formatPath(template, {
        environment: env,
        ...(component && { component })
      });

      return requireModule(environment, join(environment.cwd, path));
    });
  }

  const loader: Loader = {
    global: () => requireAllEnvs(environment.paths.global),
    component: (name) => requireAllEnvs(environment.paths.component, name)
  };
  const reducers = createReducerList(createLoaderReducers(loader, mergeValues));
  const { reduce } = createReducerExecutor(reducers);
  const environment = createEnvironment({
    cwd,
    extensions: getNodeExtensions(),
    setReducers: reducers.setReducers,
    resetReducers: reducers.resetReducers,
    reduce
  });

  return environment;
}
