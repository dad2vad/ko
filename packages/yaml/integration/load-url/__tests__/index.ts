import execa from "execa";
import { dirname, join } from "node:path";
import { installPackage } from "@kosko/test-utils";
import { createServer, Server } from "node:http";
import { AddressInfo } from "node:net";

const testDir = dirname(__dirname);
let server: Server;

function execute(env?: Record<string, string>) {
  const address = server.address() as AddressInfo;

  return execa.node(join(testDir, "index.mjs"), {
    cwd: testDir,
    env: {
      SERVER_ADDRESS: `localhost:${address.port}`,
      ...env
    }
  });
}

beforeAll(async () => {
  await installPackage(testDir, "yaml");
  server = createServer((req, res) => {
    res.writeHead(200);
    res.end(`---
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
`);
  }).listen(0);
});

afterAll(() => {
  server.close();
});

describe("when fetch is polyfilled", () => {
  test("should be able to fetch", async () => {
    const result = await execute({ NODE_OPTIONS: "-r cross-fetch/polyfill" });

    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when fetch is not polyfilled", () => {
  test("should be able to fetch", async () => {
    const result = await execute();

    expect(result.stdout).toMatchSnapshot();
  });
});
