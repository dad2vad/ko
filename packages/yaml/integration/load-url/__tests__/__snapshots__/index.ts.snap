// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when fetch is not polyfilled should be able to fetch 1`] = `
"[
  Pod { apiVersion: 'v1', kind: 'Pod', metadata: { name: 'test-pod' } }
]"
`;

exports[`when fetch is polyfilled should be able to fetch 1`] = `
"[
  Pod { apiVersion: 'v1', kind: 'Pod', metadata: { name: 'test-pod' } }
]"
`;
